### PR TITLE
[Feature], [Model]  Add node2vec random walk and model

### PR DESCRIPTION
**Pull request by Smilexuhc**
## Description
This PR implement node2vec([node2vec: Scalable Feature Learning for Networks](https://arxiv.org/abs/1607.00653) ) random walk operator and model. 
```python
import dgl
from dgl.sampling.node2vec import node2vec_randomwalk
g = dgl.graph(([0, 1, 1, 2, 3], [1, 2, 3, 0, 0]))
node2vec_randomwalk(g, [0, 1, 2, 0], p=0.25,q=4,walk_length=4)
```

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -162,7 +162,7 @@ endif(USE_CUDA)
 if(USE_CUDA)
   cuda_add_library(dgl SHARED ${DGL_SRC})
 else(USE_CUDA)
-  add_library(dgl SHARED ${DGL_SRC})
+  add_library(dgl SHARED ${DGL_SRC} src/graph/sampling/randomwalks/node2vec_impl_tmp.h)
```
Comment (BarclayII) [inline CMakeLists.txt:165]: Did you include this file in the PR?

Comment (Smilexuhc) [reply CMakeLists.txt]: It is a temporary file. I forgot to delete in the Cmakelist

File: CMakeLists.txt [hunk 2]
```diff
@@ -162,7 +162,7 @@ endif(USE_CUDA)
 if(USE_CUDA)
   cuda_add_library(dgl SHARED ${DGL_SRC})
 else(USE_CUDA)
-  add_library(dgl SHARED ${DGL_SRC})
+  add_library(dgl SHARED ${DGL_SRC} src/graph/sampling/randomwalks/node2vec_impl.h)
```
Comment (BarclayII) [inline CMakeLists.txt:165]: This file is not being included anywhere?

File: src/graph/sampling/randomwalks/node2vec_randomwalk.h [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+        int64_t i;
+        dgl_id_t curr = seed_data[seed_id];
+        dgl_id_t pre = curr;
+        traces_data[seed_id * walk_length] = curr;
+
+        for (i = 0; i < walk_length; ++i) {
+            const auto &succ = step(traces_data + seed_id * walk_length, curr, pre, i);
+            pre = curr;
+            curr = succ.first;
+            traces_data[seed_id*walk_length+i+1] = curr;
```
Comment (yzh119) [inline src/graph/sampling/randomwalks/node2vec_randomwalk.h:171]:
```suggestion
            traces_data[seed_id * walk_length + i + 1] = curr;
```

File: src/graph/sampling/randomwalks/node2vec_randomwalk.h [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+        TerminatePredicate<IdxType> terminate) {
+    const IdxType *offsets = static_cast<IdxType *>(edges[0]->data);
+    const IdxType *all_succ = static_cast<IdxType *>(edges[1]->data);
+    const IdxType *succ = all_succ + offsets[curr];
+
+    const int64_t size = offsets[curr + 1] - offsets[curr];
+
+    // Isolated node
+    if (size == 0)
+        return std::make_pair(-1, true);
```
Comment (yzh119) [inline src/graph/sampling/randomwalks/node2vec_randomwalk.h:80]: You can just return `return {-1, true};` in C++11.

File: src/graph/sampling/randomwalks/node2vec_randomwalk.h [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+    const IdxType *all_succ = static_cast<IdxType *>(edges[1]->data);
+    const IdxType *u_succ = all_succ + offsets[u];
+    const int64_t size = offsets[u + 1] - offsets[u];
+
+    return std::find(u_succ, u_succ+size, v) != u_succ+size;
+}
+
+
+template<DLDeviceType XPU, typename IdxType>
+std::pair<dgl_id_t, bool> Node2vecRandomWalkStep(
```
Comment (yzh119) [inline src/graph/sampling/randomwalks/node2vec_randomwalk.h:62]: Please write docstring to describe the arguments and return values.

File: src/graph/sampling/randomwalks/node2vec_randomwalk.h [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+bool has_edge_between(
+        const std::vector<IdArray> &edges,
+        dgl_id_t u,
+        dgl_id_t v) {
+    const IdxType *offsets = static_cast<IdxType *>(edges[0]->data);
+    const IdxType *all_succ = static_cast<IdxType *>(edges[1]->data);
+    const IdxType *u_succ = all_succ + offsets[u];
+    const int64_t size = offsets[u + 1] - offsets[u];
+
+    return std::find(u_succ, u_succ+size, v) != u_succ+size;
```
Comment (yzh119) [inline src/graph/sampling/randomwalks/node2vec_randomwalk.h:57]:
```suggestion
    return std::find(u_succ, u_succ + size, v) != u_succ + size;
```

File: src/graph/sampling/randomwalks/node2vec_randomwalk.h [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+
+namespace {
+
+
+/*!
+ * \brief Random walk step function
+ */
+template<typename IdxType>
+using Node2vecStepFunc = std::function<std::pair<dgl_id_t, bool>(
+                IdxType *,    // node IDs generated so far
```
Comment (yzh119) [inline src/graph/sampling/randomwalks/node2vec_randomwalk.h:39]: four space here.

File: src/graph/sampling/randomwalks/node2vec_randomwalk.h [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+                        break;
+                } else if (r < prob2) {
+                    break;
+                }
+            }
+        }
+    }
+    curr = next_node;
+
+    return std::make_pair(curr, terminate(data, curr, len));
```
Comment (yzh119) [inline src/graph/sampling/randomwalks/node2vec_randomwalk.h:144]: ditto.

File: src/graph/sampling/randomwalks/node2vec_randomwalk.h [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+
+#pragma omp parallel for
+    for (int64_t seed_id = 0; seed_id < num_seeds; ++seed_id) {
+        int64_t i;
+        dgl_id_t curr = seed_data[seed_id];
+        dgl_id_t pre = curr;
+        traces_data[seed_id * walk_length] = curr;
+
+        for (i = 0; i < walk_length; ++i) {
+            const auto &succ = step(traces_data + seed_id * walk_length, curr, pre, i);
```
Comment (yzh119) [inline src/graph/sampling/randomwalks/node2vec_randomwalk.h:168]: You can use `std::tie` to simplify destructing tuples.

**Review by yzh119 — COMMENTED** [review]
Some style problems.

File: src/graph/sampling/randomwalks/node2vec_randomwalk.h [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+      idx = RandomEngine::ThreadLocal()->RandInt(size);
+      next_node = succ[idx];
+    } else {
+      while (true) {
+        idx = RandomEngine::ThreadLocal()->RandInt(size);
+        r = RandomEngine::ThreadLocal()->Uniform(0., 1.);
+        next_node = succ[idx];
+        if (next_node == pre) {
+          if (r < prob0) break;
+        } else if (has_edge_between<IdxType>(edges, next_node, pre)) {
```
Comment (BarclayII) [inline src/graph/sampling/randomwalks/node2vec_randomwalk.h:108]: This is calling `has_edge_between` (i.e. `std::find`) repeatedly.  Did you compare with using `phmap::flat_hash_map` (a faster `unordered_map`)?

**Review by BarclayII — CHANGES_REQUESTED** [review]

**Comment by yzh119** [comment]
@Smilexuhc any further progress on this PR?

**Comment by BarclayII** [comment]
Closing this PR due to commit history being messed up (again; wonder why it keeps adding the master updates as separate commits...)

**Pull request closed without merging**